Allow arbitrary extern ABI

Rustfmt fails to format the following file, even though rustc happily compiles it.

I am using cfg here to repro, but the more important case in practice would be to allow non-standard ABI in the input of procedural macro attributes like https://github.com/dtolnay/cxx.

```rust
#[cfg(any())]
extern "C++" {}
```

```console
error[E0703]: invalid ABI: found `C++`
 --> /git/testing/src/main.rs:2:8
  |
2 | extern "C++" {}
  |        ^^^^^ invalid ABI
  |
  = help: valid ABIs: cdecl, stdcall, fastcall, vectorcall, thiscall, aapcs, win64, sysv64, ptx-kernel, msp430-interrupt, x86-interrupt, amdgpu-kernel, Rust, C, system, rust-intrinsic, rust-call, platform-intrinsic, unadjusted
```

I guess the error message was from the rustfmt that ships with stable 1.42, but after trying it with today's rustfmt master branch (9124dd88d6ef14d0ae77dc52ff5e9598f24a75a0) the current behavior is even worse; it silently replaces the ABI with "Rust".

```console
$ echo '#[cfg(any())] extern "C++" {}' | rustup run 1.42.0 rustfmt
error[E0703]: invalid ABI: found `C++`
 --> <stdin>:1:22
  |
1 | #[cfg(any())] extern "C++" {}
  |                      ^^^^^ invalid ABI
  |
  = help: valid ABIs: cdecl, stdcall, fastcall, vectorcall, thiscall, aapcs, win64, sysv64, ptx-kernel, msp430-interrupt, x86-interrupt, amdgpu-kernel, Rust, C, system, rust-intrinsic, rust-call, platform-intrinsic, unadjusted
```

```console
$ echo '#[cfg(any())] extern "C++" {}' | rustup run nightly-2020-03-19 rustfmt
#[cfg(any())]
extern "Rust" {}
```

```console
$ git rev-parse HEAD
9124dd88d6ef14d0ae77dc52ff5e9598f24a75a0

$ echo '#[cfg(any())] extern "C++" {}' | cargo run --bin rustfmt
#[cfg(any())]
extern "Rust" {}
```

--- a/rustfmt-core/rustfmt-lib/Cargo.toml
+++ b/rustfmt-core/rustfmt-lib/Cargo.toml
@@ -54,10 +54,6 @@ version = "642.0.0"
 package = "rustc-ap-rustc_span"
 version = "642.0.0"
 
-[dependencies.rustc_target]
-package = "rustc-ap-rustc_target"
-version = "642.0.0"
-
 [dependencies.syntax]
 package = "rustc-ap-syntax"
 version = "642.0.0"

--- a/rustfmt-core/rustfmt-lib/src/utils.rs
+++ b/rustfmt-core/rustfmt-lib/src/utils.rs
@@ -2,7 +2,6 @@ use std::borrow::Cow;
 
 use rustc_ast_pretty::pprust;
 use rustc_span::{sym, BytePos, ExpnId, Span, Symbol, SyntaxContext};
-use rustc_target::spec::abi;
 use syntax::ast::{
     self, Attribute, CrateSugar, MetaItem, MetaItemKind, NestedMetaItem, NodeId, Path, Visibility,
     VisibilityKind,

--- a/rustfmt-core/rustfmt-lib/src/utils.rs
+++ b/rustfmt-core/rustfmt-lib/src/utils.rs
@@ -140,19 +140,17 @@ pub(crate) fn format_extern(
     is_mod: bool,
 ) -> Cow<'static, str> {
     let abi = match ext {
-        ast::Extern::None => abi::Abi::Rust,
-        ast::Extern::Implicit => abi::Abi::C,
-        ast::Extern::Explicit(abi) => {
-            abi::lookup(&abi.symbol_unescaped.as_str()).unwrap_or(abi::Abi::Rust)
-        }
+        ast::Extern::None => "Rust".to_owned(),
+        ast::Extern::Implicit => "C".to_owned(),
+        ast::Extern::Explicit(abi) => abi.symbol_unescaped.to_string(),
     };
 
-    if abi == abi::Abi::Rust && !is_mod {
+    if abi == "Rust" && !is_mod {
         Cow::from("")
-    } else if abi == abi::Abi::C && !explicit_abi {
+    } else if abi == "C" && !explicit_abi {
         Cow::from("extern ")
     } else {
-        Cow::from(format!("extern {} ", abi))
+        Cow::from(format!(r#"extern "{}" "#, abi))
     }
 }
 

--- a/rustfmt-core/rustfmt-lib/tests/source/issue_4086.rs
+++ b/rustfmt-core/rustfmt-lib/tests/source/issue_4086.rs
@@ -1,0 +1,2 @@
+#[cfg(any())]
+extern  "C++"     {}

--- a/rustfmt-core/rustfmt-lib/tests/target/issue_4086.rs
+++ b/rustfmt-core/rustfmt-lib/tests/target/issue_4086.rs
@@ -1,0 +1,2 @@
+#[cfg(any())]
+extern "C++" {}

